open-apparel-registry/open-apparel-registry
# Migrate claim extended fields

## Overview

Facility claims can contain a facility_type and processing_types.
Following moving to the new taxonomy for these values, existing types
are invalid.

The following invalid values are dropped: "Other", "Boarding", "Denim
Services". All other existing values are mapped to the equivalent
taxonomy values. Extended Fields are created for the processing_types
and facility_type.

Extended fields are also created for other applicable claim fields. 

Connects #1648 

## Demo

<img width="398" alt="Screen Shot 2022-02-24 at 3 17 18 PM" src="https://user-images.githubusercontent.com/21046714/155605048-19a1e641-8ecd-4b3d-a499-1a36df313dc1.png">
<img width="404" alt="Screen Shot 2022-02-24 at 3 17 25 PM" src="https://user-images.githubusercontent.com/21046714/155605056-6878f0ea-30cf-48d8-9e8f-e3230fb00ced.png">
<img width="346" alt="Screen Shot 2022-02-24 at 3 39 33 PM" src="https://user-images.githubusercontent.com/21046714/155605079-c512a11c-caf2-4409-adb6-0cf098c2f7c1.png">

## Notes

I pulled the current `facility_production_types` and `facility_type` values from production and [created a mapping](https://docs.google.com/document/d/1qfexjxlRmdzR9_NLzG45Knc_l2-Dxhn6JEnqYyKV4eY/edit?usp=sharing) to the new system, with the assistance of the OAR team. 

Because we are filtering out / removing some invalid values, and because we need to update the actual claim object for the facility_type and processing_types fields, it wasn't possible to directly use the existing scripts for processing without a lot of edits, and it didn't make sense to rewrite them for this very specific one-off use, so I ended up copying over a lot of the code and editing where needed in the migrations file. 

As part of this migration, I checked out out an earlier version of the develop branch before the latest version of the taxonomy was applied, ran `resetdb`, and created several claims. Then I ran migrations on the current version of develop and on this branch, and confirmed that the migrations ran. 

## Testing Instructions

* Manual create old-style claims (or see notes for an optional alternate testing path).
     * Run `./scripts/server`, login, and claim two facilities. Approve the claims as an admin. Fill out the claim fields for name, native language name, parent company, number of workers, and product types. 
     * Run `./scripts/manage shell_plus`, then (use the appropriate claim ids):
```
f = FacilityClaim.objects.get(id=1)
f.facility_production_types = ['Cut and Sew / RMG', 'Boarding', 'Denim Services', 'Fusing', 'Marker Making', 'Pattern Grading', 'Sample Making', 'Trimming/Embellishment', 'Warehouse']
f.facility_type = 'Cut and Sew / RMG'
f.save()

f2 = FacilityClaim.objects.get(id=2)
f2.facility_production_types = []
f2.facility_type = 'Other'
f2.save()
```
     * Confirm in the shell that the claims match the structure of existing claims. 
* Run `./scripts/manage migrate`. 
* View the claim details and confirm that for claim 1 the new values match the expected mapping, and for claim 2 no values are present. 
* View the claimed facilities' detail pages and confirm that the extended field values match the expected taxonomy mapping. 

## Checklist
